### PR TITLE
Convert to TypeScript.

### PR DESCRIPTION
**Pull request by mythmon**
This is a big set of changes that touches every line of code. We should either do this now, or never do it. The longer we wait the harder things will get.

This totally breaks docs. I have some ideas about how to fix this, but I wanted to get feedback about the code changes before I rewrite the documentation system.

CC @andymikulski 

f?

### CODE REVIEW

**Comment by mythmon** [comment]
@rehandalal I think this is ready to go now, including docs and linting. In theory, it should auto-publish to Github pages for the doc storage. That's untested until it hits master though.

File: src/sampling.ts [hunk 1]
```diff
@@ -1,70 +1,64 @@
...
-
   /* Conceptually, this works by making a space that is the range from
    * 0 to sum(weights), assigning each choice a proportional section
    * of that space, choosing a point in that space, and returning the
    * object that corresponds to that point.
    *
    * options = [{name: "A", weight: 1}, {name: "B", weight: 2}, {name: "C", weight: 3}]
-   * totalWeight = 1 + 2 + 3 = 6
-   * choice = hashFraction(input) * totalWeight = 1.5
+   * maxWeight = 1 + 2 + 3 = 6
```
[the remark this answers is not in the document]
Comment (mythmon) [reply src/sampling.ts]: You're right, this is confusing. When I started this project, I was based on a way old commit, and I guess this snuck in. I'll put this back to the newer version.

File: src/sampling.ts [hunk 2]
```diff
@@ -1,70 +1,64 @@
...
- * @returns {number} Float between 0 and 1, inclusive.
+ * @return Float between 0 and 1, inclusive.
  */
-export async function hashFraction(input) {
+export async function hashFraction(input: string): Promise<number> {
   const hash = await sha256(input);
   const bits = 48; // meaningful precision of a 64bit floating point number
   const substringSize = bits / 4;
-  return parseInt(hash.substr(0, substringSize), 16) / Math.pow(2, bits);
+  return parseInt(hash.substr(0, substringSize), 16) / Math.pow(16, bits);
```
Comment (rehandalal) [inline src/sampling.ts]: oops also this got reverted.

File: src/sampling.ts [hunk 3]
```diff
@@ -41,30 +45,27 @@ export async function chooseWeighted(options, hashKey) {
...
- * @returns {string} Hex encoded sha256 hash.
+ * @return Hex encoded sha256 hash.
  */
-export async function sha256(message) {
+export async function sha256(message: string) {
   const msgBuffer = new TextEncoder("utf-8").encode(message);
   const hashBuffer = await crypto.subtle.digest("SHA-256", msgBuffer);
   const hashArray = Array.from(new Uint8Array(hashBuffer));
-  const hashHex = hashArray.map(b => ("00" + b.toString(16)).slice(-2)).join("");
+  const hashHex = hashArray.map((b) => ("00" + b.toString(16)).slice(-2)).join("");
```
Comment (rehandalal) [inline src/sampling.ts]: nit: extra parens
```
.map(b => (...
```

**Pull request merged**